### ISSUE TITLE
ndsctl failed to authenticate client for no apparent reason

### ISSUE THREAD

**Issue by minhng99** [openNDS/openNDS#256]
This particular client just failed to authenticate for no apparent reason, how can I debug this? After restarting `opennds` service then the issue fixed itself.

![image](https://user-images.githubusercontent.com/14113155/139518984-85227433-f3f5-4cbd-a8a4-acaf2d3ffc8b.png)

```
Sat Oct 30 00:03:31 2021 daemon.notice opennds[5928]: openNDS Version 9.3.0 is in startup
Sat Oct 30 00:03:31 2021 daemon.notice opennds[5928]: MHD version is 0.9.71
Sat Oct 30 00:03:31 2021 daemon.notice opennds[5928]: Attempting to Bind to interface: br-lan
Sat Oct 30 00:03:31 2021 daemon.notice opennds[5928]: Interface br-lan is up
Sat Oct 30 00:03:31 2021 daemon.err opennds[5928]: Could not get MAC address information of br-lan, exiting...
Sat Oct 30 00:03:36 2021 daemon.notice opennds[6207]: openNDS Version 9.3.0 is in startup
Sat Oct 30 00:03:36 2021 daemon.notice opennds[6207]: MHD version is 0.9.71
Sat Oct 30 00:03:36 2021 daemon.notice opennds[6207]: Attempting to Bind to interface: br-lan
Sat Oct 30 00:03:36 2021 daemon.notice opennds[6207]: Interface br-lan is up
Sat Oct 30 00:03:36 2021 daemon.notice opennds[6207]: Interface br-lan is at 10.15.99.1 (20:78:52:da:fb:21)
Sat Oct 30 00:03:36 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 00:03:36 2021 daemon.notice opennds[6207]: Preemptive authentication is enabled
Sat Oct 30 00:03:36 2021 daemon.notice opennds[6207]: ipset support is available
Sat Oct 30 00:03:36 2021 daemon.notice opennds[6207]: dnsmasq ipset support is available
Sat Oct 30 00:03:37 2021 daemon.notice opennds[6207]: Created web server on 10.15.99.1:2050
Sat Oct 30 00:03:37 2021 daemon.notice opennds[6207]: Maximum Html Page size is [ 10240 ] Bytes
Sat Oct 30 00:03:37 2021 daemon.notice opennds[6207]: Socket access at /tmp/ndsctl.sock
Sat Oct 30 00:03:37 2021 daemon.notice opennds[6207]: FAS Enabled.
Sat Oct 30 00:03:37 2021 daemon.notice opennds[6207]: sha256sum provider is available
Sat Oct 30 00:03:37 2021 daemon.notice opennds[6207]: SSL Provider is active
Sat Oct 30 00:03:37 2021 daemon.notice opennds[6207]: FAS URL is http://nwisp.fptplay.net:9988/v2/index.html
Sat Oct 30 00:03:37 2021 daemon.notice opennds[6207]: Forwarding Authentication is Enabled.
Sat Oct 30 00:03:37 2021 daemon.notice opennds[6207]: Initializing firewall rules
Sat Oct 30 00:03:37 2021 daemon.notice opennds[6207]: openNDS is now running.
Sat Oct 30 00:11:31 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 00:21:53 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 00:33:06 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 00:41:53 2021 daemon.notice opennds[6207]: Adding 10.15.99.165 7a:5e:d4:f2:43:be token 01cf2440 to client list
Sat Oct 30 00:43:04 2021 daemon.notice opennds[6207]: Authenticating 10.15.99.165 7a:5e:d4:f2:43:be
Sat Oct 30 00:44:05 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 00:54:21 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 00:54:26 2021 daemon.notice opennds[6207]: Adding 10.15.99.163 c0:b6:58:88:11:40 token 88c0ded6 to client list
Sat Oct 30 00:54:26 2021 daemon.notice opennds[6207]: Authenticating 10.15.99.163 c0:b6:58:88:11:40
Sat Oct 30 01:05:23 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 01:15:25 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 01:25:27 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 01:35:35 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 01:45:07 2021 daemon.notice opennds[6207]: Timeout authenticated idle user: 10.15.99.165 7a:5e:d4:f2:43:be, inactive: 3600s, in: 15kB, out: 11kB
Sat Oct 30 01:45:07 2021 daemon.notice opennds[6207]: Deauthenticating 10.15.99.165 7a:5e:d4:f2:43:be
Sat Oct 30 01:47:03 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 01:54:37 2021 daemon.notice opennds[6207]: Timeout authenticated idle user: 10.15.99.163 c0:b6:58:88:11:40, inactive: 3611s, in: 0kB, out: 0kB
Sat Oct 30 01:54:37 2021 daemon.notice opennds[6207]: Deauthenticating 10.15.99.163 c0:b6:58:88:11:40
Sat Oct 30 01:57:27 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 02:07:42 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 02:15:07 2021 daemon.notice opennds[6207]: Timeout preauthenticated idle user: 10.15.99.165 7a:5e:d4:f2:43:be, inactive: 1800s
Sat Oct 30 02:15:07 2021 daemon.notice opennds[6207]: Deleting 10.15.99.165 7a:5e:d4:f2:43:be token 9f0882da from client list
Sat Oct 30 02:18:08 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 02:24:37 2021 daemon.notice opennds[6207]: Timeout preauthenticated idle user: 10.15.99.163 c0:b6:58:88:11:40, inactive: 1800s
Sat Oct 30 02:24:37 2021 daemon.notice opennds[6207]: Deleting 10.15.99.163 c0:b6:58:88:11:40 token 2a446681 from client list
Sat Oct 30 02:28:33 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 02:29:17 2021 daemon.notice opennds[6207]: Adding 10.15.99.130 82:a0:2b:cd:cd:ef token 7560e420 to client list
Sat Oct 30 02:29:17 2021 daemon.notice opennds[6207]: Authenticating 10.15.99.130 82:a0:2b:cd:cd:ef
Sat Oct 30 02:39:55 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 02:50:06 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 03:01:29 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 03:08:11 2021 daemon.notice opennds[6207]: Adding 10.15.99.164 04:d6:aa:49:cf:04 token c9f5bf39 to client list
Sat Oct 30 03:08:11 2021 daemon.notice opennds[6207]: Authenticating 10.15.99.164 04:d6:aa:49:cf:04
Sat Oct 30 03:10:55 2021 daemon.notice opennds[6207]: Adding 169.254.131.109 34:f3:9a:36:eb:33 token 2384d0ff to client list
Sat Oct 30 03:10:55 2021 daemon.notice opennds[6207]: Authenticating 169.254.131.109 34:f3:9a:36:eb:33
Sat Oct 30 03:11:00 2021 daemon.notice opennds[6207]: Adding 10.15.99.224 34:f3:9a:36:eb:33 token f4310cc2 to client list
Sat Oct 30 03:12:51 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 03:20:24 2021 daemon.notice opennds[6207]: Adding 10.15.99.165 7a:5e:d4:f2:43:be token 113cd25e to client list
Sat Oct 30 03:20:24 2021 daemon.notice opennds[6207]: Authenticating 10.15.99.165 7a:5e:d4:f2:43:be
Sat Oct 30 03:22:55 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 03:29:37 2021 daemon.notice opennds[6207]: Timeout authenticated idle user: 10.15.99.130 82:a0:2b:cd:cd:ef, inactive: 3600s, in: 2kB, out: 1kB
Sat Oct 30 03:29:37 2021 daemon.notice opennds[6207]: Deauthenticating 10.15.99.130 82:a0:2b:cd:cd:ef
Sat Oct 30 03:33:31 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 03:36:03 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 03:36:54 2021 daemon.notice opennds[6207]: Upstream gateway [ 192.168.1.1 ] via interface [ eth0 ] is online
Sat Oct 30 03:37:39 2021 daemon.notice opennds[6207]: Authenticating 10.15.99.130 82:a0:2b:cd:cd:ef
```

**Comment by minhng99** [comment]
oh, on the closer look at the log, apparently for some reason the client has received the IP `169.254.131.109` which isn't provided by any DHCP server on the AP... which is very weird

**Comment by bluewavenet** [comment]
@minhng99 
Yes it is a Windows APIPA address.
I have seen this a few times (ie very rarely) and it was always a Windows laptop.
Reasons I found on particular machines were:
 * user (or some security/antivirus program) had blocked the entire 10.x.x.x subnet, the machine still negotiated an ip address with dhcp but then used a 169.254 address. clients.dhcp showed the allocation, but the machine showed 169.254.x.x
 * DHCP server was configured for a 200 ip address pool, the windows machine was the 201st
 
It _could_ also be the dhcp server down or very slow, but I have never seen this.

You have pre-emptive authentication enabled so your ndsctl auth should have worked if the client had just been allocated an ip address, but it failed. This would be expected if it had dropped back to APIPA as it would not be active in the arp table.

> can we make it ...... try to refresh the IP address of the client?

No, it is entirely up to the client. Telling the user to disconnect and reconnect the wifi _should_ make the client machine renegotiate dhcp but obviously if it is a client blocked subnet it will never work.

**Comment by minhng99** [comment]
I think the client already did received a valid IP address in the same range of br-lan... otherwise they would never be able to trigger the CPD and the log telling it's authenticating should never appeared, the owner of the device ask me why does it keep ask them to authenticate even though he already pressed the login button, I tried to manually run `ndsctl auth` and are getting the result above.

But how is it possible that the APIPA address show up in the opennds log? Assuming that it's using the APIPA range in that moment then it won't be able to reach the gateway

**Comment by bluewavenet** [comment]
It is also possible the client machine was in an area with very poor signal and it lost its connection and reverted to APIPA. I don't think a windows machine would automatically retry dhcp in that circumstance even if the signal improved. In this case it could well be in the arp table with the apipa address so it could indeed be authenticated by ndsctl auth [mac] if pre-emptive authentication is on, but of course it would not have access to anything except perhaps other windows machines.
Access by a foreign ip addresses is prevented in openNDS v9.5.0 (pre release currently in beta) and this should help here, but this is certainly worth investigating a little closer.

**Comment by minhng99** [comment]
> it could indeed be authenticated by ndsctl auth [mac] if pre-emptive authentication is on

Well, I did have this enabled, the client actually showing a correct IP in `ndsctl status` but somehow the `ndsctl auth [mac]` still showing failed without any additional info

**Comment by bluewavenet** [comment]
ndsctl auth [mac] would fail if the device had dropped out of the arp table again - like I said, I will investigate further.....

It is clearly not my second scenario as the client did initially get an ip address, but subsequently dropped back to apipa.
Could the machine have an intermittently faulty wireless?

**Comment by minhng99** [comment]
This person also has tried to disconnect wlan after the first failed attempt to login... and try connect again but still getting the same result so I think that the ARP table should have a good idea about this host already..

**Comment by minhng99** [comment]
![image](https://user-images.githubusercontent.com/14113155/139525301-87dd3fcc-ffc0-42a1-9994-29d22bbe38be.png)

It's happened again, trying to authenticate using the IP then it's worked

**Comment by bluewavenet** [comment]
I think it could well be a faulty wireless card in the client machine..... working, then dropping out, then working again. Windows sees dropping out as no network etc

What does `ip -f inet neigh` show?

**Comment by minhng99** [comment]
![image](https://user-images.githubusercontent.com/14113155/139525474-b55ddcf8-0112-4494-be63-0b3d0a503982.png)

**Comment by bluewavenet** [comment]
10.15.99.244 is no longer in the arp table and 169.254.131.109 is stale (ie inactive/gone away).
Probably means they have given up.
If this is only happening with this client (34:f3:9a:36:eb:33) then I really do think it is a client hardware problem.

**Comment by minhng99** [comment]
> Probably means they have given up.

Nah, they just contacted telling me that they've disconnected from wlan for some minutes then come back and the same issue happened, usually, my dnsmasq script should automatically `ndsctl auth` this client because it's assuming the client is roaming... but this failed so the CPD show up and they try to authenticate also failed

**Comment by bluewavenet** [comment]
>  If this is only happening with this client (34:f3:9a:36:eb:33) then I really do think it is a client hardware problem.

**Comment by minhng99** [comment]
Okay... okay... I'll crawl through my syslog to see if other AP having such issues....

**Comment by bluewavenet** [comment]
I have seen a Dell laptop that could reliably send wifi packets, but would regularly fail to receive for seconds at a time and windows would drop to apipa. This was unrelated to an openNDS scenario, but supports my theory.

**Comment by bluewavenet** [comment]
While you are searching logs, I will do a deep dive into the code.... :-D

**Comment by minhng99** [comment]
yeah... it looks like that there's only 2 AP out of 3k AP having this issue, it just... weird...

**Comment by bluewavenet** [comment]
@minhng99 
openNDS uses mac address as a unique client identifier - there is really no other option.
An additional test has been added to 9.5.0beta (not yet pushed to Github). 
This was added to cover the edge case where a vlan trunk is used to distribute numerous ip networks to APs with multiple SSIDs eg "guests" and "staff" where only "guests" is managed by openNDS. A client might move from guests to staff then try to access the http://status.client page. This test returns an error 403 instead of the client status page.

In your scenario this will stop the weird log entries but will not help the client.
I am feeling confident that the apipa problem is caused by one of two situations:
 1. The client wireless has a fault of some kind
 2. The client is in a fringe or shadow area with a very poor signal - or perhaps very high interference levels eg sitting next to a microwave oven.

**In your scenario, is this additional test sufficient, or should the test be refined to also check arp table entries?**

A refined test could allow the client to be genuinely authenticated but then their connection would be very poor or even broken.

**Which is best, no connection or a potentially broken/unreliable connection?**

**Comment by minhng99** [comment]
Last time this happened, I tried to authenticate using the "real" IP address instead of the MAC and it's worked, maybe you could add a test to check if the IP is invalid for the br-lan range and thus search for any other available IP for that MAC address in the ARP table?

**Comment by bluewavenet** [comment]
Yes that is what I was thinking and it should be possible.

**Comment by bluewavenet** [comment]
This was most likely caused by a faulty Windows client wireless card, that in turn made openNDS preemptive authentication fail with a "foreign" ip address. This is fixed in v9.5.0 so closing now.
Should the problem persist, a new issue should be opened.
